statediff/service.go

```diff
@@ -0,0 +1,337 @@
+			stateChanges := sds.filterByWatchedAddresses(stateChangeEvent.StateChanges)
+
+			payload, processingErr := processStateChanges(stateChanges, stateChangeEvent.Block)
+			if processingErr != nil {
+				// The service loop continues even if processing one StateChangeEvent fails
+				log.Error(fmt.Sprintf("Error processing state for block %d; error: %s ",
+					stateChangeEvent.Block.Number(), processingErr.Error()))
+			}
+
+			isEmpty, err := isEmptyPayload(payload, stateChangeEvent.Block)
+			if err != nil {
+				log.Warn("Error checking is payload is empty")
+			}
+			// Send a payload to subscribers only if isn't empty
+			if !isEmpty {
+				sds.send(payload)
+			}
+		case err := <-errCh:
+			log.Warn("Error from state change event subscription, breaking loop", "error", err)
+			sds.close()
+			return
+		case <-sds.QuitChan:
+			log.Info("Quitting the statediffing process")
+			sds.close()
+			return
+		}
+	}
+}
+
+// processStateChanges builds the state diff Payload from the modified accounts in the StateChangeEvent
+func processStateChanges(stateChanges state.StateChanges, block *types.Block) (Payload, error) {
+	var accountDiffs []AccountDiff
+	var emptyPayload Payload
+
+	// Iterate over state changes to build AccountDiffs
+	for addr, modifiedAccount := range stateChanges {
+		a, err := buildAccountDiff(addr, modifiedAccount)
+		if err != nil {
+			return emptyPayload, err
+		}
+
+		accountDiffs = append(accountDiffs, a)
+	}
+
+	stateDiff := StateDiff{
+		BlockNumber:     block.Number(),
+		BlockHash:       block.Hash(),
+		UpdatedAccounts: accountDiffs,
+	}
+
+	stateDiffRlp, err := rlp.EncodeToBytes(stateDiff)
+	if err != nil {
+		return emptyPayload, err
+	}
+	payload := Payload{
+		StateDiffRlp: stateDiffRlp,
+	}
+
+	return payload, nil
+}
+
+// buildAccountDiff
+func buildAccountDiff(addr common.Address, modifiedAccount state.ModifiedAccount) (AccountDiff, error) {
+	emptyAccountDiff := AccountDiff{}
+	accountBytes, err := rlp.EncodeToBytes(modifiedAccount.Account)
+	if err != nil {
+		return emptyAccountDiff, err
+	}
+
+	var storageDiffs []StorageDiff
+	for k, v := range modifiedAccount.Storage {
+		// Storage diff value should be an RLP object too
+		encodedValueRlp, err := rlp.EncodeToBytes(v[:])
+		if err != nil {
+			return emptyAccountDiff, err
+		}
+		storageKey := k
+		diff := StorageDiff{
+			Key:   storageKey[:],
+			Value: encodedValueRlp,
+		}
+		storageDiffs = append(storageDiffs, diff)
+	}
+
+	address := addr
+	return AccountDiff{
+		Key:     address[:],
+		Value:   accountBytes,
+		Storage: storageDiffs,
+	}, nil
+}
+
+func (sds *Service) addressInWatchedAddresses(address common.Address) bool {
+	for _, watchedAddress := range sds.WatchedAddresses {
+		checkSummedWatchedAddress := common.HexToAddress(watchedAddress).Hex()
+		if checkSummedWatchedAddress == address.Hex() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (sds *Service) filterByWatchedAddresses(stateChanges state.StateChanges) state.StateChanges {
```

Comment:
normally I'm all in favor of moving this kind of logic to a separate package (as you've done here), but it seems like this kind of work is done (for logs) in the [`filters` package](https://github.com/ethereum/go-ethereum/blob/d90d1db609c8d77baa422d49bd371207c06b4711/eth/filters/filter_system.go#L311) - thinking maybe we want to move this there, but also wondering ways we could minimize the footprint. Maybe it would make sense to introduce a new go file in that package for stuff like this and `buildAccountDiff` 🤔